### DISCUSSION TITLE
Mfront-fenics plasticity parallelization

### DISCUSSION THREAD

**Discussion by aravinth-ravikumar** [thelfer/tfel#396] [Q&A] [answered]
Dear Tfel community,

I am trying to implement anisotropic plasticity by modifying the existing plasticity code from Thomas Helfer provided [here](https://comet-fenics.readthedocs.io/en/latest/demo/plasticity_mfront/plasticity_mfront.py.html). I am having trouble running it in parallel.

A regular piece of fenics code can be run in parallel just by running the fenics python script with mpirun. But that does not work for fenics-mfront plasticity implementation. There is mismatch of sizes error when communicating functions from the mfront environment to the fenics environment. The website says "MGIS can handle parallel integration of the constitutive law which has not been used for the present computation", but I am not able to find any examples or tutorial codes on parallelization. Any help regarding this would be helpful.

**Comment by thelfer** [comment]
Dear @aravinth-ravikumar ,
Have you considered using mgis.fenics by @bleyerj  : https://thelfer.github.io/mgis/web/mgis_fenics.html
This is much easier and known to work in //

**Reply by aravinth-ravikumar** [reply]
Dear Dr. Helfer,

thank you for your reply, I will have a look at the link you have provided.

**Reply by aravinth-ravikumar** [reply]
@thelfer In the plasticity example provided using the mgis.fenics module, how can I bring the stresses and tangent operator (sig and Ct in the mfront environment) to the fenics environment? I want to do to something similar to the line 
p = problem.get_state_variable("EquivalentPlasticStrain")
but for the stresses and tangent operator Ct.

I want to do this because I want to construct my own bilinear form and residual, which requires sig and Ct. I tried using the get_state_variable() function, but I was not successful.

**Reply by thelfer** [reply]
Hi @aravinth-ravikumar,
If you want to have your own bilinear form and residual, then the only thing to do is to mimic what is done in `mgis.fenics`
Best,
Thomas

**Reply by bleyerj** [reply]
Hi,
you have something like:
```
sig = problem.get_flux("Stress")
```
This object is a specific QuadratureFunction object which possesses the following attributes:
```
sig.function.vector().get_local() # values of the stress at the quadrature points
sig.tangent_blocks # dictionary of the different tangent blocks related to the function
```

**Reply by aravinth-ravikumar** [reply]
@bleyerj Thank you very much. That worked. Can you also please tell me how to do the same for the tangent operator (the variable Dt in the mfront file)? 
By the way, are these keywords specified anywhere in a manual or are they documented anywhere?

**Reply by bleyerj** [reply]
Something like
```
sig.tangent_blocks["Strain"].function.vector().get_local()
```
should give you the dsig/deps tangent values
The "Stress" and "Strain" keyword for fluxes and gradients are default names for mechanical behaviours in MFront. For generalized behaviours, they will correspond to the names you declare when specifying the corresponding `@Flux` and `@Gradient`. See the mgis.fenics demos

**Reply by aravinth-ravikumar** [reply]
@bleyerj Thanks for your reply. I tried what you specified, but it resulted in the following error:

```
Traceback (most recent call last):
  File "mgis_plasticity.py", line 187, in <module>
    print(p1.tangent_blocks["Strain"].function.vector().get_local())
AttributeError: 'Function' object has no attribute 'tangent_blocks'
```
Where p1 = problem.get_flux("Stress")
Could you please help resolve this issue?

**Reply by bleyerj** [reply, marked as answer]
Sorry, `get_flux` returns in fact the dolfin Function. So it should be:
```
sig = problem.fluxes["Stress"]
sig_values = sig.function.vector().get_local()
Ct = sig.tangent_blocks["Strain"]
Ct_values = Ct.vector().get_local()
```

**Reply by aravinth-ravikumar** [reply]
Thank you very much. This has solved the issue.
